src/autocast_fp8.cpp

```diff
@@ -0,0 +1,77 @@
+/*
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2015-2023 Advanced Micro Devices, Inc. All rights reserved.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+#include <migraphx/autocast_fp8.hpp>
+#include <migraphx/make_op.hpp>
+#include <migraphx/program.hpp>
+#include <migraphx/instruction.hpp>
+#include <migraphx/iterator_for.hpp>
+
+namespace migraphx {
+inline namespace MIGRAPHX_INLINE_NS {
+
+void autocast_fp8_pass::apply(module& m) const
+{
+    for(auto ins : iterator_for(m))
+    {
+        const auto& ins_name = ins->name();
+        if(ins_name == "@param" && ins->get_shape().type() == fp8_type)
+        {
+            migraphx::shape new_shape = ins->get_shape().with_type(fp32_type);
+            std::string new_param_name =
+                ins->get_operator().to_value()["parameter"].to<std::string>() + "_fp32";
+            auto new_param = m.add_parameter(new_param_name, new_shape);
+            auto new_ins = m.insert_instruction(
+                ins,
+                migraphx::make_op("convert", {{"target_type", migraphx::to_value(fp8_type)}}), new_param);
+            m.replace_instruction(ins, new_ins);
+        }
+
+        if(ins_name == "@return")
+        {
+            std::vector<instruction_ref> orig_inputs = ins->inputs();
+            std::vector<instruction_ref> new_inputs;
+            bool is_input_fp8 = false;
+            for(const auto& i : orig_inputs)
+            {
+                if (i->get_shape().type() == fp8_type)
+                {
+                    is_input_fp8 = true;
```

Comment:
You can do something like this instead of keeping a flag
https://github.com/ROCm/AMDMIGraphX/blob/fe61d940c420f6e0692b90b65cb1eb3013fde312/src/eliminate_data_type.cpp#L55

Reply:
Added check for new_inputs.size() in the if condition for this.